Addons/testExoticQuadrature.hpp

```diff
@@ -0,0 +1,162 @@
+    auto sinc = [](double x) -> double {return (x == 0.0 ? 1.0 : sin(x) / (x));};
+    // Different behaviors expected for shift == 0.0 vs shift != 0.0.
+    std::vector<double> shifts = {0.0, 1.0};
+    for (auto shift : shifts) {
+        TasGrid::CustomTabulated ct = TasGrid::getExoticQuadrature(n, shift, sinc, nref, descr);
+        if (strcmp(ct.getDescription(), descr) != 0) {
+            std::cout << "ERROR: Test failed in testCtAttributes() for shift = " << shift << " on getDescription()"<< std::endl;
+            passed = false;
+        }
+        if (ct.getNumLevels() != n) {
+            std::cout << "ERROR: Test failed in testCtAttributes() for shift = " << shift << " on getNumLevels()" << std::endl;
+            passed = false;
+        }
+        for (int i=0; i<n; i++) {
+            if (ct.getQExact(i) != 2*(i+1)-1) {
+                std::cout << "ERROR: Test failed in testCtAttributes() for shift = " << shift << ", i = " << i
+                          << " on getQExact()" << std::endl;
+                passed = false;
+            }
+        }
+    }
+    return passed;
+}
+
+
+// Given a dimension, integrates the function f(x[1],...,x[dimension]) * sinc(freq * x[1]) * ... * sinc(freq * x[dimension]) over
+// [-1, 1]^dimension  using Exotic quadrature at a given depth and a specified shift. Then, compares this integral against a given exact
+// integral value.
+inline bool wrapSincTest(std::function<double(std::vector<double>)> f, int dimension, int depth, double freq, double shift,
+                         double exact_integral, double precision = 1e-10) {
+
+    // Initialize.
+    bool passed = true;
+    auto sinc = [freq](double x)->double {return (x == 0.0 ? 1.0 : sin(freq * x) / (freq * x));};
+    int level = depth % 2 == 1 ? (depth + 1) / 2 : depth / 2 + 1;
+    TasGrid::CustomTabulated ct = TasGrid::getExoticQuadrature(level, shift, sinc);
+    TasGrid::TasmanianSparseGrid sg;
+    sg.makeGlobalGrid(dimension, 1, depth, TasGrid::type_qptotal, std::move(ct));
+
+    // Compute the integral and compare to the reference value.
+    std::vector<double> quad_points = sg.getPoints();
+    std::vector<double> quad_weights = sg.getQuadratureWeights();
+    double approx_integral = 0.0;
+    assert(quad_weights.size() * dimension == quad_points.size());
+    for (size_t i=0; i<quad_weights.size(); i++) {
+        std::vector<double> x(dimension);
+        for (int d=0; d<dimension; d++) {
+            x[d] = quad_points[dimension * i + d];
+        }
+        approx_integral += f(x) * quad_weights[i];
+    }
+    if (std::abs(approx_integral - exact_integral) > precision) {
+        std::cout << "ERROR: " << std::setprecision(16) << "Computed integral value " << approx_integral
+                  << " does not match exact integral " << exact_integral << " for test problem " << std::endl
+                  << "∫_[-1,1]^n f(x[1],...,x[n]) * sinc(freq*x[n]) * ... * sinc(freq*x[n]) dx[1] ... dx[n]" << std::endl
```

Comment:
This is very verbose, but OK. As a general note, when you debug this it is better to know which `f` failed rather than the general form of the integral. You can get the `f` from the dimension, so it's OK here, otherwise we would need another variable that is some form of a description.